core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -272,15 +364,270 @@ class RemoteLogManager(rlmConfig: RemoteLogManagerConfig,
+          // LSO indicates the offset below are ready to be consumed(high-watermark or committed)
+          val lso = log.lastStableOffset
+          if (lso < 0) {
+            warn(s"lastStableOffset for partition $tpId is $lso, which should not be negative.")
+          } else if (lso > 0 && copiedOffset < lso) {
+            // Copy segments only till the min of high-watermark or stable-offset as remote storage should contain
+            // only committed/acked messages
+            val toOffset = lso
+            debug(s"Checking for segments to copy, copiedOffset: $copiedOffset and toOffset: $toOffset")
+            val activeSegBaseOffset = log.activeSegment.baseOffset
+            // log-start-offset can be ahead of the read-offset, when:
+            // 1) log-start-offset gets incremented via delete-records API (or)
+            // 2) enabling the remote log for the first time
+            val fromOffset = Math.max(copiedOffset + 1, log.logStartOffset)
+            val sortedSegments = log.logSegments(fromOffset, toOffset).toSeq.sortBy(_.baseOffset)
+            val activeSegIndex: Int = sortedSegments.map(x => x.baseOffset).search(activeSegBaseOffset) match {
+              case Found(x) => x
+              case InsertionPoint(y) => y - 1
+            }
+            // sortedSegments becomes empty list when fromOffset and toOffset are same, and activeSegIndex becomes -1
+            if (activeSegIndex < 0) {
+              debug(s"No segments found to be copied for partition $tpId with copiedOffset: $copiedOffset and " +
+                s"active segment's base-offset: $activeSegBaseOffset")
+            } else {
+              sortedSegments.slice(0, activeSegIndex).foreach { segment =>
+                if (isCancelled() || !isLeader()) {
+                  info(s"Skipping copying log segments as the current task state is changed, cancelled: " +
+                    s"${isCancelled()} leader:${isLeader()}")
+                  return
+                }
+
+                val logFile = segment.log.file()
+                val logFileName = logFile.getName
+
+                info(s"Copying $logFileName to remote storage.")
+                val id = new RemoteLogSegmentId(tpId, Uuid.randomUuid())
+
+                val nextOffset = segment.readNextOffset
+                val endOffset = nextOffset - 1
+                val producerIdSnapshotFile: File = log.producerStateManager.fetchSnapshot(nextOffset).orElse(null)
+
+                val epochEntries = getLeaderEpochCheckpoint(log, segment.baseOffset, nextOffset).read()
+                val segmentLeaderEpochs: util.Map[Integer, lang.Long] = new util.HashMap[Integer, lang.Long](epochEntries.size())
+                epochEntries.forEach(entry =>
+                  segmentLeaderEpochs.put(Integer.valueOf(entry.epoch), lang.Long.valueOf(entry.startOffset))
+                )
+
+                val remoteLogSegmentMetadata = new RemoteLogSegmentMetadata(id, segment.baseOffset, endOffset,
```

Comment:
could we rename this variable to a meaningful name like `copySegmentStartedMetadata`?